### PR TITLE
Refactor CompilerSupplier

### PR DESCRIPTION
**Pull request by gnidan**
This PR makes changes to the CompilerSupplier's interface and to its implementation.

## Interface changes

This PR strips out extra methods defined for CompilerSupplier, as well as removes the need to invoke individual LoadingStrategies directly.

Instead of `supplier.downloadAndCacheSolc()`,`supplier.getDockerTags()`, etc., the CompilerSupplier now exposes:
- `await supplier.load()`
   **Note**: method used to return `{ solc, parserSolc }`. It now returns only `{ solc }` and the parser-determination logic has been moved to the one place it was used (`loadParser()`).
  (Taking a line in the sand: the CompilerSupplier is easy. You construct one and then do `await supplier.load()` and you get supplied with one compiler.)
- `await supplier.list()`
  This steals the existing return-type of `{ latestRelease, releases, prereleases }` from `versionRange.getReleases()`, but implements it for the Docker strategy as well.
  **Caveat here**: Docker's API returns paginated results, so `releases` and `prereleases` are both now AsyncIterableIterators instead of regular arrays. `truffe compile --list=docker --all` now works, but is slow, and prints a warning not to use it.

## Implementation changes

- Move compiler-caching functionality is now in its own class, used by Docker and VersionRange strategies
- Move the LoadingStrategy base class's listener management methods to a single function, `observeListeners`, which returns an interface for cleaning up listeners (`const listeners = observeListeners(); /* ... */; listeners.cleanup();`)
- Use custom error classes instead of `this.errors(kind, ...)` lookup, moving common error classes to an errors module.
- Stop relying on LoadingStrategy's base class constructor, which was handling all strategies' options together. Individual strategies now define their own constructor, asking for what they need.
- Upon doing all of this, notice that the LoadingStrategy base class is now empty and thus remove it.

### CODE REVIEW

File: packages/compile-solidity/compilerSupplier/Cache.js
Comment (eggplantzzz) [inline packages/compile-solidity/compilerSupplier/Cache.js:11]: Didn't we just cache the compilers in the `compilers` directory? Do we put it in `node_modules` because that is ignored by babel processing when it comes to using `require` on a path in `node_modules`?

Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/Cache.js:11]: I think it's for us... we have a package or two that uses babel, and IIRC it explodes without this.

Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/Cache.js:11]: (this has been this way; I haven't changed this)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (haltman-at) [inline packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:195]: This whole function appears to be rolling its own retry logic.  It'd probably be better to tear that out and just use `axios-retry` like you're doing elsewhere.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (haltman-at) [inline packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:158]: This appears to be another function using its own retry logic, for which I'd once again suggest using `axios-retry` instead.

Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:158]: Also, I know this is preexisting, but the name of this function is confusing.  I'd suggest renaming it to `getSolcByUrlAndCacheIt` to make it clearer.  (As it is, it sounds like it takes a URL and a cache as input and returns a solc as output, when in fact it takes just a url as input, returns a solc as output, and also caches that solc.)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (haltman-at) [inline packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:138]: This is a preexisting, I know, but... it looks like this function isn't used anywhere.  Can it be deleted?

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js
Comment (haltman-at) [inline packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js:55]: I will as always suggest using `Boolean(x)` instead of `!!x`. :)

**Review by haltman-at — COMMENTED** [review]
Whew, OK, got through all this.  I have two more comments that GitHub wouldn't let me add, because they weren't close enough to any code that was changed:

1. In `compilerSupplier/index.js`, line 18 looks like a mistake.  If it's not a mistake, it needs a comment for sure, but I'm guessing it's just a mistake.  Possibly fixing that should be a separate PR.

2. In `VersionRange.js`, the `findNewestValidVersion` function is unnecessary, because the `semver` library already has the `maxSatisfying` function.  You might want to replace the one with the other, as long as you're doing all this.

...actually, I'd like to just generally note that I was really hoping to see the giant tangle that is `VersionRange.js` cut through and reduced to something more sensible, but, eh, I guess that's not what you were trying to do.  Oh well, perhaps some other time.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/LoadingStrategy.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/LoadingStrategy.js:33]: (Seconded!)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:15]: Oh, yeah, I didn't notice that...

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:55]: That's what #4225 does :)

File: packages/core/test/lib/commands/compile.js
Comment (gnidan) [reply packages/core/test/lib/commands/compile.js:118]: It used to print 11 items. Don't know why. But Docker gives back 10 items per page, so I switched to that

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js:74]: They're AsyncIterableIterators, so the idea is to consume via `for await (const release of releases) { ... }`. This works for both strategies.

Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js:74]: (#4225 will add types for this)

File: packages/compile-solidity/compilerSupplier/index.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/index.js:55]: Oh, cool!

File: packages/core/test/lib/commands/compile.js
Comment (haltman-at) [reply packages/core/test/lib/commands/compile.js:118]: Huh, OK.

File: packages/compile-solidity/compileWithPragmaAnalysis.js
Comment (haltman-at) [reply packages/compile-solidity/compileWithPragmaAnalysis.js:72]: Heh, I guess I *didn't* understand why it was there after all! :)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:15]: It redirects (302 Found) to the [now outdated] Solidity URL for these. It's there in case solc's own distribution methods fail; we can switch to hosting the compilers ourselves at this URL.

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:11]: Will open docs PR to add those

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:13]: Hm, it's a bit tricky here, because the default is necessary for the compiler supplier's interface, and I don't want to introduce the @truffle/config dependency. Note how this doesn't actually take a config (it takes that `{ solcConfig, ... }` options object)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Local.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Local.js:8]: Ah perhaps not. I was just leaving all the internal implementation the same and concerning myself with the interface. I'll clean it up.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js:19]: Hm, a mystery. No idea.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:17]: Probably separate PR... I'll make an issue for it

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:138]: I also noticed that. I wonder the provenance of this. I can eliminate.

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:118]: In this file, that condition won't happen - it uses either VersionRange or Docker only. Inside the implementation of `supplier.list()`, there's a check.

It's a bit weird, but I've addressed this in #4224 by making it explicit whether a strategy supplies listing functionality; or see #4225 for implementation.

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:120]: Oh yeah totally. Not sure why this got renamed; I just didn't bother to change it. Will do so.

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:145]: Not sure? The reason this is here is because previously there was no way to list `--all` for Docker. I've changed that, so now `truffle compile --list=docker --all` will work, but it'll print a warning to note that the operation is quite slow. (The difference is: solc gives us a single JSON response with all versions, but Docker paginates their results... and there are >500 pages of results)

File: packages/compile-solidity/compilerSupplier/index.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/index.js:13]: Note I have no idea if changing this is actually a good idea; I think leaving it alone is fine!  I don't see how this would introduce a dependency though?  Like how would putting a default version in the config defaults introduce a dependency here?

Anyway yeah probably makes sense to leave this alone for now, or at least to address it in a separate PR.

File: packages/core/lib/commands/compile.js
Comment (haltman-at) [reply packages/core/lib/commands/compile.js:145]: Hm, whatever I guess!

File: packages/core/lib/commands/compile.js
Comment (haltman-at) [reply packages/core/lib/commands/compile.js:118]: It looks to me like that condition can happen?  If the solc version is set to `"native"` or a local soljson (and `docker` is not set to true), it looks like the CompilerSupplier will be created with the corresponding loading strategy.  What am I missing?

I do see that the check you mentioned exists, so that we at least throw a proper error, but this should be a can't-happen and right now it looks like it isn't?

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (cds-amal) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:15]: That's important context! Can you add it as a comment please?

**Comment by cds-amal** [comment]
I think it would be better git practice to get the interface refactor merged and then rework the implementation in followup work.

File: packages/compile-solidity/compilerSupplier/Cache.js
Comment (eggplantzzz) [reply packages/compile-solidity/compilerSupplier/Cache.js:11]: oh I thought this was new, I think the issue is that babel processes all required items and the compiler is so big that it will explode trying to process it all

File: packages/compile-solidity/compilerSupplier/index.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/index.js:13]: @eggplantzzz, I don't understand the question.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Local.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Local.js:4]: thanks! I thought of you when doing that one ;)

File: packages/compile-solidity/compilerSupplier/loadingStrategies/LoadingStrategy.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/LoadingStrategy.js:33]: heckin' swell

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js:55]: Let's see what [my Twitter poll](https://twitter.com/gnidan/status/1425960578049265674) says

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:11]: I'll start with [this issue](https://github.com/trufflesuite/trufflesuite.com/issues/1043)

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:100]: I propose to fix this in #4225 to avoid a needless merge conflict

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:78]: This whole class is getting eliminated in #4225; making an error class here is thus unnecessary?

File: packages/compile-solidity/compilerSupplier/index.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/index.js:89]: I added [a note](https://github.com/trufflesuite/truffle/pull/4225/files#r688150977) to do this in #4225.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js:19]: Mind if we leave this as-is for now, since it's unclear? I can open an issue to track it down if you think it's a good idea.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:22]: Issue opened [here](https://github.com/trufflesuite/truffle/issues/4252)

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:117]: Added [a note](https://github.com/trufflesuite/truffle/pull/4225/files#r688157235) to ensure this gets addressed in #4225.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:200]: Opened [an issue](https://github.com/trufflesuite/truffle/issues/4261) for this

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:163]: > This appears to be another function using its own retry logic, for which I'd once again suggest using `axios-retry` instead.

Opened [issue](https://github.com/trufflesuite/truffle/issues/4261) for this

Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:163]: > Also, I know this is preexisting, but the name of this function is confusing.

Renamed it to `getAndCacheSolcByUrl`

File: packages/compile-solidity/compilerSupplier/Cache.js
Comment (gnidan) [reply packages/compile-solidity/compilerSupplier/Cache.js:11]: oh cool, good to know!

File: packages/compile-solidity/compilerSupplier/index.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/index.js:78]: Oh, OK then.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Native.js:19]: Yup, just leaving this alone for now seems fine.

File: packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js
Comment (haltman-at) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/Docker.js:63]: :P

File: packages/core/lib/commands/compile.js
Comment (haltman-at) [reply packages/core/lib/commands/compile.js:117]: I think that it would be better to address this here since the problem is being introduced here?

Comment (cds-amal) [reply packages/core/lib/commands/compile.js:117]: I agree with Harry. If the problem is introduced in this PR it should be fixed here.

**Comment by gnidan** [comment]
@haltman-at for your other comments:
 
> 1. In `compilerSupplier/index.js`, line 18 looks like a mistake.  If it's not a mistake, it needs a comment for sure, but I'm guessing it's just a mistake.  Possibly fixing that should be a separate PR.

#4225 removes this class (and thus removes the problem).

> 2. In `VersionRange.js`, the `findNewestValidVersion` function is unnecessary, because the `semver` library already has the `maxSatisfying` function.  You might want to replace the one with the other, as long as you're doing all this.

Hm, could we revisit this later? I just tried this and it's not as clean as you might expect, since the `allVersions` variable needs to be massaged to match the input for `semver.maxSatisfying`. Want I should add a note to #4225?

> ...actually, I'd like to just generally note that I was really hoping to see the giant tangle that is `VersionRange.js` cut through and reduced to something more sensible, but, eh, I guess that's not what you were trying to do. Oh well, perhaps some other time.

Sorry :(

Thanks for taking the time to review so thoughtfully!

File: packages/core/lib/commands/compile.js
Comment (gnidan) [reply packages/core/lib/commands/compile.js:117]: Sound reasoning! One sec :)

Comment (gnidan) [reply packages/core/lib/commands/compile.js:121]: OK pushed. It's not as clean as my solution offered in #4225, because in this PR, Docker still requires a version. (In the other PR, I fixed it to conform.)

Let me know if this is alright!

**Comment by gnidan** [comment]
@eggplantzzz @cds-amal I got your comments; let me know if there's anything else you happen to spot!

Thanks both of you ❤️

File: packages/core/lib/commands/compile.js
Comment (haltman-at) [reply packages/core/lib/commands/compile.js:121]: The hacky solution is fine by me.

**Review by haltman-at — APPROVED** [review]
Cool, I think this is good for now!

**Review by cds-amal — APPROVED** [review]
Looks good to me. CompilerSupplier is in a better place :)

**Comment by gnidan** [comment]
Thanks 🙇 

I'm gonna wait for the other PRs before merging this... or should I make a feature branch?

... I wanna hit the merge button; I'm gonna make a feature branch. 😁

File: packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js
Comment (cameel) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:22]: https://ethereum.github.io/solc-bin/bin/ is deprecated but it's just not receiving new binaries. Old ones are still there. Keeping it here should be fine for the time being and could actually provide an additional fallback in case the other options go down. It would only be a problem if we needed to rebuild one of the binaries and replace it but these cases are rare (currently there's just one that might end up being rebuilt eventually: https://github.com/ethereum/solidity/issues/11671).

Also, you might want to add https://solc-bin.ethereum.org/wasm/ (or rather https://binaries.soliditylang.org/wasm/ but it's just a cosmetic difference as they both point at the same thing) to the list above `bin/`. It has wasm rather than asm.js builds and uses the same naming convention so it can be used as the primary source with `bin/` being a fallback. Wasm builds offer better performance and don't have problems like https://github.com/ethereum/solidity/issues/11671. In 0.6.2 we switched to doing wasm builds and putting them in `bin/` but older binaries were not replaced. Instead wasm rebuilds of older versions are in `wasm/`.

To make this more apparent there are new dirs that contain the same binaries but correctly separated (`emscripten-asmjs/` and `emscripten-wasm32/`). I'd recommend using them if possible, though unfortunately they are not drop-in replacements since the platform name is now a part of the naming scheme.

Note also that each directory, including `wasm/`, has a separate `list.json` and wasm binaries obviously have different hashes than the asm.js ones so maybe it does not fit into this system after all. At least if you do not download `list.json` from the same entry you use to download binaries.

Comment (cds-amal) [reply packages/compile-solidity/compilerSupplier/loadingStrategies/VersionRange.js:22]: Thanks @cameel ! 

cc: @gnidan

**Pull request merged**
